src/cluster/mcl.cpp

```diff
@@ -0,0 +1,570 @@
+							if(it.row() == it.col()){
+								attractors.emplace(it.row());
+							}
+						}
+					}
+					list_of_sets = disjointSet.getListOfSets();
+#ifdef MCL_TIMINGS
+					sparse_list_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+				}
+				else{
+					n_dense++;
+					Eigen::MatrixXf m_dense = Eigen::MatrixXf::Zero(order.size(), order.size());
+					for(Eigen::Triplet<float> const & t : m){
+						m_dense(t.row(), t.col()) = t.value();
+					}
+#ifdef MCL_TIMINGS
+					dense_create_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+					markov_process(&m_dense, inflation, expansion);
+#ifdef MCL_TIMINGS
+					chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
+#endif
+					LazyDisjointIntegralSet<uint32_t> disjointSet(m_dense.cols());
+					for (uint32_t icol=0; icol<m_dense.cols(); ++icol){
+						for (uint32_t irow=0; irow<m_dense.rows(); ++irow){
+							if( abs(m_dense(irow, icol)) > numeric_limits<float>::epsilon()){
+								disjointSet.merge(irow, icol);
+								if(irow == icol){
+									attractors.emplace(irow);
+								}
+							}
+						}
+					}
+					list_of_sets = disjointSet.getListOfSets();
+#ifdef MCL_TIMINGS
+					dense_list_time += chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t).count();
+#endif
+				}
+				for(unordered_set<uint32_t> subset : list_of_sets){
+					assert(cluster_id < 0x00ffffff);
+					for(uint32_t el : subset){
+						const uint32_t mask = attractors.find(el) == attractors.end() ? MASK_NORMAL_NODE : MASK_ATTRACTOR_NODE;
+						clustering_result[order[el]] = mask | cluster_id;
+					}
+					if(subset.size() == 1) n_singletons++;
+					cluster_id += nThreads;
+				}
+			}
+			else if (order.size() == 1){
+				assert(cluster_id < 0x00ffffff);
+				clustering_result[order[0]] = MASK_SINGLE_NODE | cluster_id;
+				cluster_id += nThreads;
+				n_singletons++;
+			}
+			//my_counter = component_counter.fetch_add(1, memory_order_relaxed);
+		}
+		n_clusters_found.fetch_add((cluster_id-iThr)/nThreads, memory_order_relaxed);
+		n_dense_calculations.fetch_add(n_dense, memory_order_relaxed);
+		n_sparse_calculations.fetch_add(n_sparse, memory_order_relaxed);
+		nClustersEq1.fetch_add(n_singletons, memory_order_relaxed);
+	};
+
+	vector<thread> threads;
+	for(uint32_t iThread = 0; iThread < nThreads ; iThread++ ) {
+		min_sparsities[iThread] = 1.0f;
+		max_sparsities[iThread] = 0.0f;
+		threads.emplace_back(mcl_clustering, iThread);
+	}
+
+	float min_sparsity = 1.0f;
+	float max_sparsity = 0.0f;
+	for(uint32_t iThread = 0; iThread < nThreads ; iThread++ ) {
+		threads[iThread].join();
+		min_sparsity = min(min_sparsity, min_sparsities[iThread]);
+		max_sparsity = max(max_sparsity, max_sparsities[iThread]);
+	}
+	delete[] min_sparsities;
+	delete[] max_sparsities;
+
+	timer.finish();
+	cout << "Found "
```

Comment:
`cout` should be reserved for the TSV output that goes into the file. You can use `message_stream` for this.

Reply:
done.